### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by tkarna**
Adding docstrings to all classes/functions. Adding math to all equation docstrings.

Addresses some points of #76

### CODE REVIEW

**Comment by tkarna** [comment]
I've started by writing documentation to `shallowwater_eq.py`, `rungekutta.py` and `callback.py`. If these look all right I'll do the same for the rest.

I dumped temporary html files here:
http://www.stccmop.org/~karnat/tmp/thetis_www/thetis.html#module-thetis.shallowwater_eq
http://www.stccmop.org/~karnat/tmp/thetis_www/thetis.html#module-thetis.rungekutta
http://www.stccmop.org/~karnat/tmp/thetis_www/thetis.html#module-thetis.callback

**Comment by tkarna** [comment]
AFAIK, this is good to go now. Please review if you feel it's needed.

File: thetis/coupled_timeintegrator.py [hunk 1]
```diff
@@ -328,7 +356,21 @@ def _create_integrators(self):
...
-        # TODO decide which dt to use where ...
+        """
+        Set time step for the coupled time integrator
+
+        :arg dt: Time step. This is the master (macro) time step used to
+            march the 3D equations.
+        :type dt: float
+        :arg dt_2d: Time step for 2D equations. For constency :attr:`dt_2d`
+            must be an integer fraction of :attr:`dt`. If 2D solver is implicit
+            use set :attr:`dt_2d` equal to :attr:`dt`.
```
Comment (wence-) [inline thetis/coupled_timeintegrator.py]: `s/use//`

File: thetis/coupled_timeintegrator.py [hunk 2]
```diff
@@ -421,10 +473,11 @@ def advance(self, t, dt, update_forcings=None, update_forcings3d=None):
 class CoupledERKALE(CoupledTimeIntegrator):
     """
     Implicit-Explicit SSP RK solver for conservative ALE formulation
+
```
Comment (wence-) [inline thetis/coupled_timeintegrator.py]: Just "Explicit SSP RK"?  The next sentence says it's fully explicit.

Comment (wence-) [inline thetis/coupled_timeintegrator.py]: Oh, I see the vertical integrator is backward euler.

File: thetis/equation.py [hunk 3]
```diff
@@ -9,8 +9,15 @@
...
     """
     Implements a single term of an equation.
+
+    .. note::
+        Sign convention: all terms are assumed to be on the left hand side of the equation A + term = 0.
     """
     def __init__(self, function_space):
+        """
+        :arg function_space: :class:`FunctionSpace` where the solution belongs
+            to
```
Comment (wence-) [inline thetis/equation.py]: `s/the//`

File: thetis/equation.py [hunk 4]
```diff
@@ -62,6 +65,10 @@ class Equation(object):
     SUPPORTED_LABELS = ['source', 'explicit', 'implicit', 'nonlinear']
 
     def __init__(self, function_space):
+        """
+        :arg function_space: :class:`FunctionSpace` where the solution belongs
```
Comment (wence-) [inline thetis/equation.py]: likewise here `s/the//`

Comment (wence-) [inline thetis/equation.py]: sorry, nonsense, "the FunctionSpace the solution belongs to" (or "the FunctionSpace where the solution belongs")

File: thetis/equation.py [hunk 5]
```diff
@@ -75,13 +82,21 @@ def __init__(self, function_space):
...
         return inner(solution, self.test) * dx
 
     def add_term(self, term, label):
+        """
+        Adds a term in the equation
+
+        :arg term: :class:`.Term` object to add_term
+        :arg string label: Assign a label to the term. Generally one of
+            'implicit', 'explicit', 'source' indicating how this term should be
+            treated in the time integrator.
```
Comment (wence-) [inline thetis/equation.py]: Could link to `SUPPORTED_LABELS`: "Valid labels are given by :attr:`SUPPORTED_LABELS`".

File: thetis/equation.py [hunk 6]
```diff
@@ -62,6 +65,10 @@ class Equation(object):
     SUPPORTED_LABELS = ['source', 'explicit', 'implicit', 'nonlinear']
```
Comment (wence-) [inline thetis/equation.py]: Can document these class members by adding a docstring below, I think.  Perhaps.

```
"""
Valid labels for terms, indicating how they should be treated in the time integrator.

source
   The term is a source term

explicit
    The term should be treated explicitly

implicit
    The term should be treated implicitly

nonlinear
    The term is nonlinear, and should be treated as such...
"""
```

File: thetis/implicitexplicit.py [hunk 7]
```diff
@@ -55,21 +56,42 @@ def get_final_solution(self):
...
 class IMEXGeneric(TimeIntegrator):
     """
-    Generic implementation of Runge-Kutta Impicit-Explicit schemes
+    Generic implementation of Runge-Kutta Implicit-Explicit schemes
+
+    Derived classes must define the implicit :attr:`dirk_class` and explicit
+    :attr:`erk_class` Runge-Kutta time integrator classes.
+
+    This method solves the linearized equations: All implicit terms are fed to
+    the implicit solver, while all the other terms are fed to the implicit
```
Comment (wence-) [inline thetis/implicitexplicit.py]: second implicit should be explicit?

File: thetis/momentum_eq.py [hunk 8]
```diff
@@ -1,12 +1,74 @@
...
+Above :math:`r` denotes the baroclinic head
+
+.. math::
+    r = \frac{1}{\rho_0} \int_{z}^\eta  \rho' d\zeta.
+    :label: baroc_head
+
+In the case of purely barotropic problems the :math:`r` and the internal pressure
+gradient are omitted.
+
+When using mode spliting we split the velocity field to depth average and a
```
Comment (wence-) [inline thetis/momentum_eq.py]: "into a depth average and a deviation"

Comment (wence-) [inline thetis/momentum_eq.py]: Also "splitting".

File: thetis/momentum_eq.py [hunk 9]
```diff
@@ -115,6 +214,26 @@ def residual(self, solution, solution_old, fields, fields_old, bnd_conditions=No
...
+        + \int_{\mathcal{I}_h \cup \mathcal{I}_v} \textbf{u}^{\text{up}} \cdot
+        \text{jump}(\boldsymbol{\psi} \textbf{n}_h) \cdot \text{avg}(\textbf{u}) dS
+
+    where the right hand side has been integrated by parts; :math:`:` stand for
+    the Frobenius inner product, :math:`\textbf{n}_h` is the horizontal
+    projection of the normal vector, :math:`\textbf{u}^{\text{up}}` is the
+    upwind value, and :math:`\text{jump}` and :math:`\text{avg}` denote the
+    jump and average operators across the interface.
+    If :math:`\bar{\textbf{u}}` belongs to a discontinuous function space, the
+    latter form is used.
```
Comment (wence-) [inline thetis/momentum_eq.py]: I'm not sure I understand this sentence.  Do you mean that if u_bar is discontinuous, it is upwinded?

Comment (wence-) [inline thetis/momentum_eq.py]: Oh, wait, there is an equality here, so you use the second version for discontinuous fields.

File: thetis/momentum_eq.py [hunk 10]
```diff
@@ -408,9 +626,29 @@ def residual(self, solution, solution_old, fields, fields_old, bnd_conditions=No
 
 
 class SourceTerm(MomentumTerm):
+    r"""
+    Generic momentun source term
```
Comment (wence-) [inline thetis/momentum_eq.py]: momentum.

File: thetis/coupled_timeintegrator.py [hunk 11]
```diff
@@ -328,7 +356,21 @@ def _create_integrators(self):
...
 
     def set_dt(self, dt, dt_2d):
-        # TODO decide which dt to use where ...
+        """
+        Set time step for the coupled time integrator
+
+        :arg dt: Time step. This is the master (macro) time step used to
+            march the 3D equations.
+        :type dt: float
+        :arg dt_2d: Time step for 2D equations. For constency :attr:`dt_2d`
```
Comment (wence-) [inline thetis/coupled_timeintegrator.py]: consistency?

File: thetis/equation.py (hunk 4 above)
Comment (tkarna) [inline thetis/equation.py]: thanks, fixed these.

File: thetis/equation.py (hunk 5 above)
Comment (tkarna) [inline thetis/equation.py]: Good idea, added.

File: thetis/equation.py (hunk 6 above)
Comment (tkarna) [inline thetis/equation.py]: Likewise, added.

File: thetis/implicitexplicit.py (hunk 7 above)
Comment (tkarna) [inline thetis/implicitexplicit.py]: yes, fixed.

File: thetis/momentum_eq.py (hunk 8 above)
Comment (tkarna) [inline thetis/momentum_eq.py]: yes, fixed.

File: thetis/momentum_eq.py (hunk 9 above)
Comment (tkarna) [inline thetis/momentum_eq.py]: Yes that's not entirely clear...
In fact the term only implements the RHS form so I've removed this sentence.

File: thetis/momentum_eq.py (hunk 10 above)
Comment (tkarna) [inline thetis/momentum_eq.py]: I'm so dyslexic it's not funny :D

File: thetis/coupled_timeintegrator.py (hunk 11 above)
Comment (tkarna) [inline thetis/coupled_timeintegrator.py]: fixed.

File: thetis/equation.py (hunk 6 above)
Comment (tkarna) [inline thetis/equation.py]: It seems `nonlinear` is not actually used anywhere... Could potentially remove that.

**Comment by tkarna** [comment]
Merci bien! Merging ...

**Pull request merged**